#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
data_path = '/content/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [2]:
df.groupby('Branch')['gross income'].sum().sort_values(ascending = False)

,gross income
Branch,
C,5265.1765
A,5057.1605
B,5057.0320


Найвищий сумарний прибуток у філіалі С (5265.1765)

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [3]:
df.groupby(['Branch', 'City'])['gross income'].sum().sort_values(ascending = False)

,,gross income
Branch,City,
C,Naypyitaw,5265.1765
A,Yangon,5057.1605
B,Mandalay,5057.0320


Філіал С знаходиться у місті Naypyitaw та приносить найбільший прибуток у розмірі 5265.1765.

Проте, аналізуючи прибуток кожного філіалу, и бачимо, що він майже однаковий.
Тобто відкривати ще одни магазин не є доцільним, адже це збільшить операційне навантаження, поточні клієнти можуть розконцентрувати свою увагу та перейти до нового агазину.


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [4]:
invoices_by_category = pd.pivot_table(df, index= 'Branch', columns= 'Product line', values= 'Invoice ID', aggfunc='count' )

In [5]:
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [6]:
invoices_by_category['Electronic accessories'].max()

60

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1. Чи відрізняється гендерна структура покупців та середній чек залежно від міста ? Хто купує? Необхідно зрозуміти, чи варто таргетувати рекламу на конкретну аудиторію у кожному міст.

2. Що саме вони купують? Які саме категорії товарів найбільше купує цей цільовий сегмент і як на це впливає наявність картки лояльності. Проаналізувавши які товари є популярними серед наших топових клієнтів, зможемо оптимізувати процес їхньої оплати.

3. Як вони за це платять? Яким способом оплати користуються ці ключові клієнти при купівлі своїх фаворитних категорій і який спосіб дає найбільший чек?

1. Чи відрізняється гендерна структура покупців та середній чек залежно від міста ? Хто купує? Необхідно зрозуміти, чи варто таргетувати рекламу на конкретну аудиторію у кожному міст.

In [7]:
city_gender_avg_total = pd.pivot_table(df, index= 'City', columns= 'Gender', values= 'Total', aggfunc= 'mean').round(2)
city_gender_avg_total

Gender,Female,Male
City,,
Mandalay,326.72,313.35
Naypyitaw,346.55,325.89
Yangon,330.86,295.71


Провівши аналіз, бачимо, що жінки найбільше витрачають у місті Naypyitaw (346,55),найменше у місті Mandalay (326,72),  а чоловіки показують найвищий чек у місті Naypyitaw (325,89) і найнижчий у Yangon (295,71).

Оскільки місто Naypyitaw є лідером за купівельною спроможністю для обох статей, тут варто просувати більш дорогі та преміальні товари.

Оскільки чоловіки у місті Yangon мають найнижчий середній чек (295.71) а жінки купують значно більше, то акції та промо-пропозиції у цьому супермаркеті варто першочергово орієнтувати на жіночу аудиторію.

2. Оскільки ми з'ясували, що жінки у місті Naypyitaw є нашим найбільш високочековим та найціннішим сегментом, тепер досліджуємо, що саме вони купують? Які саме категорії товарів найбільше купує цей цільовий сегмент і як на це впливає наявність картки лояльності. Проаналізувавши які товари є популярними серед наших топових клієнтів, зможемо оптимізувати процес їхньої оплати.

In [8]:
naypyitaw_women = df[(df['City'] == 'Naypyitaw') & (df['Gender'] == 'Female')]
naypyitaw_women.head(2)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.82,80.22,3/8/2019,10:29,Cash,76.4,4.761905,3.82,9.6
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.78,772.38,2/24/2019,11:38,Ewallet,735.6,4.761905,36.78,8.0


In [9]:
products_by_customer_type = naypyitaw_women.groupby(['Product line', 'Customer type'])['Invoice ID'].count().unstack().fillna(0)
products_by_customer_type['Total'] = products_by_customer_type.sum(axis=1)
products_by_customer_type.loc['Total'] = products_by_customer_type.sum(axis=0)
products_by_customer_type

Customer type,Member,Normal,Total
Product line,,,
Electronic accessories,9,19,28
Fashion accessories,19,16,35
Food and beverages,23,15,38
Health and beauty,11,12,23
Home and lifestyle,19,6,25
Sports and travel,15,14,29
Total,96,82,178


Food and beverages (38 покупок) та Fashion accessories (35 покупок) є абсолютними фаворитами серед жінок у Naypyitaw.

Учасники програми лояльності купують товари категорії Home and lifestyle в тричі частіше (19/6). Food and beverages також показує суттєву перевагу серед лояльних клієнтів (23/15). Electronic accessories-лише 9 покупок роблять Member,коли звичайні покупці Normal купують їх удвічі частіше (19 покупок).

Відштовхуючтсь від результатів анадізу, можемо спробувати впровадити персоналізовані знижки та підвищений кешбек за карткою Member саме на категорії Food and beverages та Home and lifestyle. А для покупців без картки, тобто Normal, які купують Electronic accessories, можемо впровадити акцію на касі та запропонувати оформити картку лояльності прямо на касі та отримати бонусні бали на наступну покупку.

3. Як вони за це платять? Яким способом оплати користуються ці ключові клієнти при купівлі своїх фаворитних категорій і який спосіб дає найбільший чек?

In [10]:
top_categories = ['Food and beverages', 'Fashion accessories']

In [11]:
naypyitaw_women_prod_line_top= df[(df['City'] == 'Naypyitaw') & (df['Gender'] == 'Female')& (df['Product line'].isin(top_categories))]
naypyitaw_women_prod_line_top.head(2)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
34,183-56-6882,C,Naypyitaw,Member,Female,Food and beverages,99.42,4,19.884,417.564,2/6/2019,10:42,Ewallet,397.68,4.761905,19.884,7.5
43,228-96-1411,C,Naypyitaw,Member,Female,Food and beverages,98.70,8,39.480,829.080,3/4/2019,20:39,Cash,789.60,4.761905,39.480,7.6


In [12]:
naypyitaw_women_prod_line_top.pivot_table(index='Product line', columns='Payment', values='Total', aggfunc='mean').round(2)

Payment,Cash,Credit card,Ewallet
Product line,,,
Fashion accessories,266.99,345.34,349.60
Food and beverages,442.42,317.45,451.41


У Fashion accessories електронні гаманці та картки показують чек майже на 80 вищий, ніж готівка, а саме 349.60 і 345.34 проти 266.99.

У Food and beverages електронні гаманці дають найвищий чек взагалі серед усіх 451.41.

Також спостерігаємо, що у категорії продуктів готівка також показує високий чек 442.42,при цьому розрахунки кредитною карткою тут значно відстають 317.45.

3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.

Результати аналізу дають змогу перейти до прийняття рішень на основі даних.
Ми з'ясували, що найщедріші покупці є жінки в Naypyitaw.Націлюємо рекламний бюджет і спеціальні акції саме на них у цьому регіоні.

Топові категорії цього сегмента -це продукти напої та аксесуари, а товари для дому гарно підходять до програми лояльності.Робимо готові набори чи пропозиції (наприклад купити аксесуар та отримати знижку на товари для дому за карткою Member")

Аналіз показав, що коли жінки купують продукти та аксесуари через Ewallet, їхній середній чек найвищий. Отже, плануємо партнерський кешбек із платіжними сервісами.